dbms/src/Storages/DeltaMerge/VersionChain/VersionChain.cpp

```diff
@@ -0,0 +1,383 @@
+        snapshot.detailInfo(),
+        curr_replayed_rows,
+        curr_replayed_deletes,
+        sw_total.elapsed() / 1000000);
+
+    return base_versions;
+}
+
+template <ExtraHandleType HandleType>
+template <typename Iter>
+void VersionChain<HandleType>::replayHandles(
+    const DMContext & dm_context,
+    Iter begin,
+    Iter end,
+    const UInt32 stable_rows,
+    DeltaValueReader & delta_reader)
+{
+    for (auto itr = begin; itr != end; ++itr)
+    {
+        const auto h = *itr;
+        if (auto row_id = new_handle_to_row_ids.find(h, delta_reader, stable_rows); row_id)
+        {
+            base_versions->push_back(*row_id);
+            continue;
+        }
+        if (auto row_id = findBaseVersionFromDMFileOrDeleteRangeList(dm_context, h); row_id)
+        {
+            base_versions->push_back(*row_id);
+            continue;
+        }
+        const RowID curr_row_id = base_versions->size() + stable_rows;
+        new_handle_to_row_ids.insert(h, curr_row_id);
+        base_versions->push_back(NotExistRowID);
+    }
+}
+
+template <ExtraHandleType HandleType>
+UInt32 VersionChain<HandleType>::replayBlock(
+    const DMContext & dm_context,
+    const IColumnFileDataProviderPtr & data_provider,
+    const ColumnFile & cf,
+    const UInt32 offset,
+    const UInt32 stable_rows,
+    const bool calculate_read_packs,
+    DeltaValueReader & delta_reader)
+{
+    assert(cf.isInMemoryFile() || cf.isTinyFile());
+
+    auto cf_reader = cf.getReader(dm_context, data_provider, getHandleColumnDefinesPtr<HandleType>(), ReadTag::MVCC);
+    auto block = cf_reader->readNextBlock();
+    RUNTIME_CHECK_MSG(
+        cf.getRows() == block.rows(),
+        "ColumnFile<{}> returns {} rows. Read all rows in one block is required!",
+        cf.toString(),
+        block.rows());
+
+    const auto & column = *(block.begin()->column);
+    RUNTIME_CHECK(column.size() > offset, column.size(), offset);
+
+    const auto handle_col = ColumnView<HandleType>(column);
+    auto itr = handle_col.begin() + offset;
+
+    if (calculate_read_packs)
+        calculateReadPacks(itr, handle_col.end());
+
+    replayHandles(dm_context, itr, handle_col.end(), stable_rows, delta_reader);
+    return column.size() - offset;
+}
+
+template <ExtraHandleType HandleType>
+UInt32 VersionChain<HandleType>::replayColumnFileBig(
+    const DMContext & dm_context,
+    const ColumnFileBig & cf_big,
+    const UInt32 stable_rows,
+    const StableValueSpace::Snapshot & stable,
+    const std::span<const ColumnFilePtr> preceding_cfs,
+    DeltaValueReader & delta_reader)
+{
+    auto cf_big_min_max = loadDMFileHandleRange<HandleType>(dm_context, *(cf_big.getFile()));
+    if (!cf_big_min_max) // DMFile is empty.
+        return 0;
+
+    HandleRefType cf_big_min = cf_big_min_max->first;
+    HandleRefType cf_big_max = cf_big_min_max->second;
+
+    // If a ColumnFileBig is ingested by apply snapshot, its preceding should be a ColumnFileDeleteRange.
```

Comment:
Is it the first phase in `ingestDTFilesUsingSplit`?

Reply:
...And if we use `ingestDTFilesUsingColumnFile`, it is in `MemTableSet::ingestColumnFiles` where we will append `ColumnFileDeleteRange`?